README
			Pathneck (version 1.2)
    	   [the copyright note is at the end of this file]

Quick Start 
===========

	$ make / make sun / make bsd
	[use "make" for Linux & MAC, use "make -f Makefile.sun" for Solaris, 
	 and "make -f Makefile.bsd" for BSD. After this step, you should get 
	 pathneck & get-choke now.]

	$ su
	[change to root]

	# ./setroot.sh 
	[set root id for pathneck, so that a normal user can run it]

	$ exit 
	[go back to normal user]

	$ ./pathneck -o <dst>

Pathneck
========
The output from "pathneck -h" is as the following:

----------------------------------------------------------------------
Pathenck V1.1 Usage: 

    	./pathneck [-e end_pkt_num] [-l udp_pkt_num] [-s pkt_size] 	
		   [-i self_ip] [-y delay_num] [-coptx] [-dvh]		
		   <dst_ip | dst_hostname>				

[probing configuration]
	-e end_pkt_num	number of measurement packets [30]		
	-l udp_pkt_num	number of load packets [60]			
	-s pkt_size	the load packet size in byte [500]		
	-c		use ICMP probing packets, [UDP]			
    	-p 		use the planetlab raw socket interface [0]	
	-y delay_num	specify the src gap within the packet train [0]	
	-i self_ip	sproof the probing pkt source ip [not set]	

[output setting] 
	-x		enable the DNS lookup [0]			
	-o 		enable on-line detection processing [0]		
	-t		dump the packet sending times [0]		
	-d 		debug mode [0]					
	-v 		verbose mode [0]				
	-h 		print this message [0]				
----------------------------------------------------------------------
    Here the "[]" denotes the default setting for the options.

    Pathneck needs root permission to send out raw packets.  setroot.sh 
    (needs root to run) can set the owner id of pathneck with "root", 
    then you will be able to run it with a normal user account

    NOTE1: The "-l" is critical for the tool to work correctly, the value 
    depends on the access link capacity and the probing host CPU speed.  
    Currently, the setup is emphirical. Some typical value:
	o. 100Mbps access link and 400MHz CPU: use the default value
	o. otherwise, eg, 10Mbps access link, or < 400MHz CPU: use "-l 20"

    NOTE2:
    The current version will only detect bottleneck for a path where we can 
    detect at least 4 gap values.

Example
-------

    o. ./pathneck -l 20 -cx yahoo.com
       
       use a 20-packet probing train, probing packets are ICMP ECHO packet,
       and enable hostname lookup

    o. ./pathneck yahoo.com

       the configuration uses the default values: using 500B UDP probing 
       packets, 60-packet probing train, will not do hostname look up.

    o.  $ ./pathneck -xo 81.161.208.1

    [The output is the following:]

    1102870265.142626 81.161.208.1 500 60 0

    00   5.512   128.2.223.254    250   2989 . 959.612 ub GIGROUTER.NET.CS.CMU.EDU
    01   1.518      128.2.0.12   2989   2989 .  80.293 lb CORE0-VL1000.GW.CMU.NET
    02   4.385    128.2.33.225    380   2989 . 629.933 lb HYPER-VL501.GW.CMU.NET
    03   1.630  192.88.115.181   2748   2748 .  87.336 lb bar-cmu-ge-4-0-0-0.psc.net
    04   1.878    192.88.115.5   2742   2748 .  87.526 lb minime-ge-0-1-0-0.3rox.net
    05   1.991  199.239.216.13   2757   2748 .  87.049 lb pos4-1-1-0.a01.pitbpa05.us.ra.verio.net
    06   2.366  129.250.24.142   2379   2748 . 100.845 lb ge-0-3-0-4.r01.pitbpa05.us.bb.verio.net
    07  11.980   129.250.5.123   3125   3125 .  76.778 lb p16-1-1-2.r21.nycmny01.us.bb.verio.net
    08  11.862    129.250.3.49   3237   3125 .  74.121 lb p16-7-0-0.r04.nycmny01.us.bb.verio.net
    09  11.851 209.244.160.149   3118   3125 .  76.948 lb so-10-1.core2.NewYork1.Level3.net
    10  11.716       4.68.97.1   3252   3125 .  73.800 lb ae-1-51.bbr1.NewYork1.Level3.net
    11  81.421    4.68.128.105   2995   3125 .  80.133 lb 4.68.128.105
    12  93.031  212.187.128.29   3249   3125 .  73.849 lb so-3-0-0.mp2.Frankfurt1.Level3.net
    13  93.277  195.122.136.39   3123   3125 .  76.848 lb ge-11-0.ipcolo1.Frankfurt1.Level3.net
    14 135.246     62.67.32.66   8994   8994 2  26.684 ub 62.67.32.66
    15 132.498   194.12.224.10   9489   8994 .  25.292 lb 194.12.224.10
    16 157.113  194.12.224.226   5102   8994 .  47.039 lb rvn4fa0-0-6.lirex.net
    17 148.108    194.12.255.2  14098  14098 3  17.023 ub lirex-m-net.lirex.net
    18 155.665 193.110.223.249  72633  72633 1   3.304 ub fw-e0.mnet.bg
    conf = 0.806 0.653 0.362

    rtt = 167.972 ( 81.161.208.1 )

    [Reading:]
    ----------
    The 1st line: 
    	probing_start_timestamp (local time), dst_ip, load_pkt_size, 
	load_pkt_num, delay_num

    The main body (the 7-column thing):
    1st column:   the index number
    2nd column:   the RTT to that hop
    3rd column:	  hop IP address
    4th column:	  raw gap measurement (in microsecond)
    5th column:	  "smooth"ed gap values (in microsecond), they are also 
    		  the input for the dynamic algo
    6th column:   choke points, "." represents non-choke point, "1", "2",
    		  "3" refers to the choke points (output at most 3), ordered
		  by the confidence value.
    7th column:   bandwidth bound estimation, unit: Mbps
    8th column:   ub->upper bound, lb->lower bound, uk->unknown (see the
	          pathneck paper section 2.3.3 for the detail.)
    9th column:   hop hostname

    The "conf" line:
    	the confidence values for the choke points detected, in the order of
	"1", "2", "3".

    The "rtt" line:
    	the RTT to the destination, dst_ip, and the hop count *from* dst to
	source (i.e., return path, not forward path).

    [Options:]
    ----------
    "$ ./pathneck -o 81.161.208.1" will not output the last column 
    "$ ./pathneck 81.161.208.1" will only output the first 4 columns

Scripts
=======

The script/ directory includes the sCRIPTS THat implements the algorithm used
in SIGCOMM'04 paper. They are mainly useful for large scale measurement:

	$ ./pathneck [-o] yahoo.com (10 times) >> tmp.txt
	$ ./script/get.sh <root_dir> tmp.txt > tmp.sum

	[a complete example script is in script/run.sh. Before using run.sh,
	make sure the "root" variable in this file is set correctly. ]

When doing large scale measurement, "-o" can increase the CPU load of the
probing nodes, and thus is not recommended. The script/get.pl (used by
script/get.sh) has the off-line processing code to run the same dynamic
algorithm to detect choke points.

Example
-------

    $ ./run.sh 81.161.208.1

    [This command will create a 2 files: 81.161.208.1.txt & 81.161.208.1.sum.
     The beginning part of 81.161.208.1.sum looks something like:]

     sum: 0.88 08 07  67161 18 193.110.223.249
     sum: 0.50 08 04   6495 14 62.67.32.66

     info 1102870355.254499 81.161.208.1 500 60 0

     623   3125        5.301 128.2.223.254        0  76.800 ub 128.2.223.254
     3125  3125        1.427 128.2.0.12           0  76.800 lb 128.2.0.12
     1502  3125        3.043 128.2.33.225         0  76.800 lb 128.2.33.225
     3746  3242        0.924 192.88.115.181       0  74.028 lb 192.88.115.181
     3242  3242        1.166 192.88.115.5         0  74.028 lb 192.88.115.5
     3380  3380        1.283 199.239.216.13       0  71.006 lb 199.239.216.13
     3500  3380        1.159 129.250.24.142       0  71.006 lb 129.250.24.142
     3114  3121       11.901 129.250.5.123        0  76.898 lb 129.250.5.123
     3121  3121       11.894 129.250.3.49         0  76.898 lb 129.250.3.49
     3123  3121       11.761 209.244.160.149      0  76.898 lb 209.244.160.149
     3119  3119       11.759 4.68.97.1            0  76.948 lb 4.68.97.1
     2995  3119       81.332 4.68.128.105         0  76.948 lb 4.68.128.105
     3247  3119       92.817 212.187.128.29       0  76.948 lb 212.187.128.29
     3002  3119       93.312 195.122.136.39       0  76.948 lb 195.122.136.39
     3998  3998  [3] 141.280 62.67.32.66          0  60.030 ub 62.67.32.66
     7994  3998      132.906 194.12.224.10        0  60.030 lb 194.12.224.10
     874   874   [2] 165.380 194.12.224.226       0 274.600 uk 194.12.224.226
     372   874       164.132 194.12.255.2         0 274.600 lb 194.12.255.2
     65958 65958 (1) 164.503 193.110.223.249      0   3.639 ub 193.110.223.249
     conf = 0.987 3.574 0.220

     rtt = 184.383 ( 81.161.208.1 )

    ... (followed by 9 more "info ... rtt" probing bodies)

    [Reading:]
    ----------
    The "sum:" lines shows the choke points detected based on the 10 probings,
    ordered by their position on the path. The first line corresponds to the 
    bottleneck.

    The "sum:" line has the following format:

    sum: [d_rate] [#appear] [#detected] [avg_gap] [hop#] [ip]
	
    Here 
    	d_rate 	= #detected/#appear
      	avg_gap is the average gap values measured in the 10 probings
	hop# 	is the position of corresponding router on the path

    NOTE: the summary info are based the "dominant route", i.e., the route
    which appears the most number of times during the 10 probings. If the
    dominant route appear less than 5 times, we do not compute any summary 
    info.

    The following probing bodies is pretty similar with those output from
    pathneck's raw probing, except that the (raw_gap_value, post_processed_gap,
    and the choke point label) is at the first 3 columns.

    [MISC:]
    -------
    o. there is no difference between "[]" & "()"
    o. the different output format between run.sh and pathneck is simply 
       because I run out of time to clean up the script code. Hopefully I can 
       do it in near future.
    o. the RTT measurement doesn't work on planetlab --- haven't figure out 
       the reason for not sending ICMP ECHO packet. Let me know if you know
       what I am doing wrong.

 
Contact
=======

Please send any question, comment, and bug report to 
Ningning Hu (hnn@cs.cmu.edu).


/****************************************************************************
 *  Pathneck: locating network path bottlenecks
 *  Copyright (C) 2004
 *  Ningning Hu and the Carnegie Mellon University
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License (in the COPYING file) for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 ****************************************************************************/


ChangeLog
Sun Dec 12 12:01:15 EST 2004

	* add in the code for bandwidth bound estimation (see README for the 
	  example). The following code are changed: 
	  choke.c, get-choke.c, pathneck.c, script/get.pl

Sat Nov  6 15:41:09 EST 2004

	* fix the bug in "-c" mode
	   [thanks Bryant Lai for the bug report]

	* release version 1.2

Thu Nov  4 19:23:32 EST 2004

	* apply the patch for MAC 
	   [thanks Michael Abd-El-Malek for the patch 
	    & Neil Spring for the testing]

	* fix the bug that can show up under "-o" mode
	* adjust the output when no measurement is obtained 
	   [thanks Neil Spring for pointing them out]

Fri Aug 27 20:26:39 EDT 2004
	
	* release version 1.1


ip_icmp.h
/* Copyright (C) 1991, 92, 93, 95, 96, 97, 99 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
   02111-1307 USA.  */

#ifndef __NETINET_IP_ICMP_H
#define __NETINET_IP_ICMP_H    1

#include <sys/cdefs.h>
//#include <sys/types.h>

__BEGIN_DECLS

struct icmphdr
{
  u_int8_t type;		/* message type */
  u_int8_t code;		/* type sub-code */
  u_int16_t checksum;
  union
  {
    struct
    {
      u_int16_t	id;
      u_int16_t	sequence;
    } echo;			/* echo datagram */
    u_int32_t	gateway;	/* gateway address */
    struct
    {
      u_int16_t	__unused;
      u_int16_t	mtu;
    } frag;			/* path mtu discovery */
  } un;
};

#define ICMP_ECHOREPLY		0	/* Echo Reply			*/
#define ICMP_DEST_UNREACH	3	/* Destination Unreachable	*/
#define ICMP_SOURCE_QUENCH	4	/* Source Quench		*/
#define ICMP_REDIRECT		5	/* Redirect (change route)	*/
#define ICMP_ECHO		8	/* Echo Request			*/
#define ICMP_TIME_EXCEEDED	11	/* Time Exceeded		*/
#define ICMP_PARAMETERPROB	12	/* Parameter Problem		*/
#define ICMP_TIMESTAMP		13	/* Timestamp Request		*/
#define ICMP_TIMESTAMPREPLY	14	/* Timestamp Reply		*/
#define ICMP_INFO_REQUEST	15	/* Information Request		*/
#define ICMP_INFO_REPLY		16	/* Information Reply		*/
#define ICMP_ADDRESS		17	/* Address Mask Request		*/
#define ICMP_ADDRESSREPLY	18	/* Address Mask Reply		*/
#define NR_ICMP_TYPES		18


/* Codes for UNREACH. */
#define ICMP_NET_UNREACH	0	/* Network Unreachable		*/
#define ICMP_HOST_UNREACH	1	/* Host Unreachable		*/
#define ICMP_PROT_UNREACH	2	/* Protocol Unreachable		*/
#define ICMP_PORT_UNREACH	3	/* Port Unreachable		*/
#define ICMP_FRAG_NEEDED	4	/* Fragmentation Needed/DF set	*/
#define ICMP_SR_FAILED		5	/* Source Route failed		*/
#define ICMP_NET_UNKNOWN	6
#define ICMP_HOST_UNKNOWN	7
#define ICMP_HOST_ISOLATED	8
#define ICMP_NET_ANO		9
#define ICMP_HOST_ANO		10
#define ICMP_NET_UNR_TOS	11
#define ICMP_HOST_UNR_TOS	12
#define ICMP_PKT_FILTERED	13	/* Packet filtered */
#define ICMP_PREC_VIOLATION	14	/* Precedence violation */
#define ICMP_PREC_CUTOFF	15	/* Precedence cut off */
#define NR_ICMP_UNREACH		15	/* instead of hardcoding immediate value */

/* Codes for REDIRECT. */
#define ICMP_REDIR_NET		0	/* Redirect Net			*/
#define ICMP_REDIR_HOST		1	/* Redirect Host		*/
#define ICMP_REDIR_NETTOS	2	/* Redirect Net for TOS		*/
#define ICMP_REDIR_HOSTTOS	3	/* Redirect Host for TOS	*/

/* Codes for TIME_EXCEEDED. */
#define ICMP_EXC_TTL		0	/* TTL count exceeded		*/
#define ICMP_EXC_FRAGTIME	1	/* Fragment Reass time exceeded	*/


#if defined __USE_BSD || defined __CYGWIN__
/*
 * Copyright (c) 1982, 1986, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	@(#)ip_icmp.h	8.1 (Berkeley) 6/10/93
 */

#include <netinet/in.h>
#include <netinet/ip.h>

/*
 * Internal of an ICMP Router Advertisement
 */
struct icmp_ra_addr
{
  u_int32_t ira_addr;
  u_int32_t ira_preference;
};

#if 0
struct ip {
#ifdef _IP_VHL
        u_char  ip_vhl;                 /* version << 4 | header length >> 2 */
#else
#if BYTE_ORDER == LITTLE_ENDIAN
        u_int   ip_hl:4,                /* header length */
                ip_v:4;                 /* version */
#endif
#if BYTE_ORDER == BIG_ENDIAN
        u_int   ip_v:4,                 /* version */
                ip_hl:4;                /* header length */
#endif
#endif /* not _IP_VHL */
        u_char  ip_tos;                 /* type of service */
        u_short ip_len;                 /* total length */
        u_short ip_id;                  /* identification */
        u_short ip_off;                 /* fragment offset field */
#define IP_RF 0x8000                    /* reserved fragment flag */
#define IP_DF 0x4000                    /* dont fragment flag */
#define IP_MF 0x2000                    /* more fragments flag */
#define IP_OFFMASK 0x1fff               /* mask for fragmenting bits */
        u_char  ip_ttl;                 /* time to live */
        u_char  ip_p;                   /* protocol */
        u_short ip_sum;                 /* checksum */
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
};
#endif

struct icmp
{
  u_int8_t  icmp_type;	/* type of message, see below */
  u_int8_t  icmp_code;	/* type sub code */
  u_int16_t icmp_cksum;	/* ones complement checksum of struct */
  union
  {
    u_char ih_pptr;		/* ICMP_PARAMPROB */
    struct in_addr ih_gwaddr;	/* gateway address */
    struct ih_idseq		/* echo datagram */
    {
      u_int16_t icd_id;
      u_int16_t icd_seq;
    } ih_idseq;
    u_int32_t ih_void;

    /* ICMP_UNREACH_NEEDFRAG -- Path MTU Discovery (RFC1191) */
    struct ih_pmtu
    {
      u_int16_t ipm_void;
      u_int16_t ipm_nextmtu;
    } ih_pmtu;

    struct ih_rtradv
    {
      u_int8_t irt_num_addrs;
      u_int8_t irt_wpa;
      u_int16_t irt_lifetime;
    } ih_rtradv;
  } icmp_hun;
#define	icmp_pptr	icmp_hun.ih_pptr
#define	icmp_gwaddr	icmp_hun.ih_gwaddr
#define	icmp_id		icmp_hun.ih_idseq.icd_id
#define	icmp_seq	icmp_hun.ih_idseq.icd_seq
#define	icmp_void	icmp_hun.ih_void
#define	icmp_pmvoid	icmp_hun.ih_pmtu.ipm_void
#define	icmp_nextmtu	icmp_hun.ih_pmtu.ipm_nextmtu
#define	icmp_num_addrs	icmp_hun.ih_rtradv.irt_num_addrs
#define	icmp_wpa	icmp_hun.ih_rtradv.irt_wpa
#define	icmp_lifetime	icmp_hun.ih_rtradv.irt_lifetime
  union
  {
    struct
    {
      u_int32_t its_otime;
      u_int32_t its_rtime;
      u_int32_t its_ttime;
    } id_ts;
    struct
    {
      struct ip idi_ip;
      /* options and then 64 bits of data */
    } id_ip;
    struct icmp_ra_addr id_radv;
    u_int32_t   id_mask;
    u_int8_t    id_data[1];
  } icmp_dun;
#define	icmp_otime	icmp_dun.id_ts.its_otime
#define	icmp_rtime	icmp_dun.id_ts.its_rtime
#define	icmp_ttime	icmp_dun.id_ts.its_ttime
#define	icmp_ip		icmp_dun.id_ip.idi_ip
#define	icmp_radv	icmp_dun.id_radv
#define	icmp_mask	icmp_dun.id_mask
#define	icmp_data	icmp_dun.id_data
};

/*
 * Lower bounds on packet lengths for various types.
 * For the error advice packets must first insure that the
 * packet is large enough to contain the returned ip header.
 * Only then can we do the check to see if 64 bits of packet
 * data have been returned, since we need to check the returned
 * ip header length.
 */
#define	ICMP_MINLEN	8				/* abs minimum */
#define	ICMP_TSLEN	(8 + 3 * sizeof (n_time))	/* timestamp */
#define	ICMP_MASKLEN	12				/* address mask */
#define	ICMP_ADVLENMIN	(8 + sizeof (struct ip) + 8)	/* min */
#ifndef _IP_VHL
#define	ICMP_ADVLEN(p)	(8 + ((p)->icmp_ip.ip_hl << 2) + 8)
	/* N.B.: must separately check that ip_hl >= 5 */
#else
#define	ICMP_ADVLEN(p)	(8 + (IP_VHL_HL((p)->icmp_ip.ip_vhl) << 2) + 8)
	/* N.B.: must separately check that header length >= 5 */
#endif

/* Definition of type and code fields. */
/* defined above: ICMP_ECHOREPLY, ICMP_REDIRECT, ICMP_ECHO */
#define	ICMP_UNREACH		3		/* dest unreachable, codes: */
#define	ICMP_SOURCEQUENCH	4		/* packet lost, slow down */
#define	ICMP_ROUTERADVERT	9		/* router advertisement */
#define	ICMP_ROUTERSOLICIT	10		/* router solicitation */
#define	ICMP_TIMXCEED		11		/* time exceeded, code: */
#define	ICMP_PARAMPROB		12		/* ip header bad */
#define	ICMP_TSTAMP		13		/* timestamp request */
#define	ICMP_TSTAMPREPLY	14		/* timestamp reply */
#define	ICMP_IREQ		15		/* information request */
#define	ICMP_IREQREPLY		16		/* information reply */
#define	ICMP_MASKREQ		17		/* address mask request */
#define	ICMP_MASKREPLY		18		/* address mask reply */

#define	ICMP_MAXTYPE		18

/* UNREACH codes */
#define	ICMP_UNREACH_NET	        0	/* bad net */
#define	ICMP_UNREACH_HOST	        1	/* bad host */
#define	ICMP_UNREACH_PROTOCOL	        2	/* bad protocol */
#define	ICMP_UNREACH_PORT	        3	/* bad port */
#define	ICMP_UNREACH_NEEDFRAG	        4	/* IP_DF caused drop */
#define	ICMP_UNREACH_SRCFAIL	        5	/* src route failed */
#define	ICMP_UNREACH_NET_UNKNOWN        6	/* unknown net */
#define	ICMP_UNREACH_HOST_UNKNOWN       7	/* unknown host */
#define	ICMP_UNREACH_ISOLATED	        8	/* src host isolated */
#define	ICMP_UNREACH_NET_PROHIB	        9	/* net denied */
#define	ICMP_UNREACH_HOST_PROHIB        10	/* host denied */
#define	ICMP_UNREACH_TOSNET	        11	/* bad tos for net */
#define	ICMP_UNREACH_TOSHOST	        12	/* bad tos for host */
#define	ICMP_UNREACH_FILTER_PROHIB      13	/* admin prohib */
#define	ICMP_UNREACH_HOST_PRECEDENCE    14	/* host prec vio. */
#define	ICMP_UNREACH_PRECEDENCE_CUTOFF  15	/* prec cutoff */

/* REDIRECT codes */
#define	ICMP_REDIRECT_NET	0		/* for network */
#define	ICMP_REDIRECT_HOST	1		/* for host */
#define	ICMP_REDIRECT_TOSNET	2		/* for tos and net */
#define	ICMP_REDIRECT_TOSHOST	3		/* for tos and host */

/* TIMEXCEED codes */
#define	ICMP_TIMXCEED_INTRANS	0		/* ttl==0 in transit */
#define	ICMP_TIMXCEED_REASS	1		/* ttl==0 in reass */

/* PARAMPROB code */
#define	ICMP_PARAMPROB_OPTABSENT 1		/* req. opt. absent */

#define	ICMP_INFOTYPE(type) \
	((type) == ICMP_ECHOREPLY || (type) == ICMP_ECHO || \
	(type) == ICMP_ROUTERADVERT || (type) == ICMP_ROUTERSOLICIT || \
	(type) == ICMP_TSTAMP || (type) == ICMP_TSTAMPREPLY || \
	(type) == ICMP_IREQ || (type) == ICMP_IREQREPLY || \
	(type) == ICMP_MASKREQ || (type) == ICMP_MASKREPLY)

#endif /* __USE_BSD */

__END_DECLS

#endif /* netinet/ip_icmp.h */
